evm/src/fuzz/invariant/executor.rs

```diff
@@ -0,0 +1,290 @@
+                            for invariant in invariants.iter() {
+                                failures
+                                    .borrow_mut()
+                                    .failed_invariants
+                                    .insert(invariant.name.clone(), Some(error.clone()));
+                            }
+
+                            break 'fuzz_run
+                        }
+                    }
+
+                    // Generates the next call from the run using the recently updated
+                    // dictionary.
+                    inputs.extend(
+                        strat
+                            .new_tree(&mut branch_runner.borrow_mut())
+                            .map_err(|_| TestCaseError::Fail("Could not generate case".into()))?
+                            .current(),
+                    );
+                }
+
+                // Before each run, we must reset the database state.
+                executor.borrow_mut().backend_mut().db = clean_db.clone();
+
+                // We clear all the targeted contracts created during this run.
+                if !created_contracts.is_empty() {
+                    let mut writable_targeted = targeted_contracts.write();
+                    for addr in created_contracts.iter() {
+                        writable_targeted.remove(addr);
+                    }
+                }
+
+                fuzz_cases.borrow_mut().push(FuzzedCases::new(fuzz_runs));
+
+                Ok(())
+            });
+        }
+
+        // TODO: only saving one sequence case per invariant failure. Do we want more?
+        let (reverts, invariants) = failures.into_inner().into_inner();
+
+        Ok(Some(InvariantFuzzTestResult { invariants, cases: fuzz_cases.into_inner(), reverts }))
+    }
+
+    /// Prepares certain structures to execute the invariant tests:
+    /// * Fuzz dictionary
+    /// * Targeted contracts
+    /// * Invariant Strategy
+    fn prepare_fuzzing(
+        &mut self,
+        invariant_address: Address,
+        test_contract_abi: &Abi,
+        test_options: InvariantTestOptions,
+    ) -> eyre::Result<(EvmFuzzState, FuzzRunIdentifiedContracts, BoxedStrategy<Vec<BasicTxDetails>>)>
+    {
+        // Finds out the chosen deployed contracts and/or senders.
+        let (targeted_senders, targeted_contracts) =
+            self.select_contracts_and_senders(invariant_address, test_contract_abi)?;
+
+        if targeted_contracts.is_empty() {
+            eyre::bail!("No contracts to fuzz.");
+        }
+
+        // Stores fuzz state for use with [fuzz_calldata_from_state].
+        let fuzz_state: EvmFuzzState = build_initial_state(&self.executor.backend().db);
+
+        // During execution, any newly created contract is added here and used through the rest of
+        // the fuzz run.
+        let targeted_contracts: FuzzRunIdentifiedContracts =
+            Arc::new(RwLock::new(targeted_contracts));
```

Comment:
why do we need a lock here?

Reply:
`targeted_contracts` needs to be accessible from inside the `Executor` Inspector for using the `CallOverrideStrategy`. It also needs to be changed (albeit from outside) on `collect_created_contracts`.